### PR TITLE
Jdbc table schema discovery

### PR DESCRIPTION
**Pull request by VardhanThigle**
Adding code for JDBC schema discovery from information system tables.

### CODE REVIEW

File: v2/sourcedb-to-spanner/src/main/java/com/google/cloud/teleport/v2/source/reader/io/schema/SchemaDiscovery.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+package com.google.cloud.teleport.v2.source.reader.io.schema;
+
+import com.google.cloud.teleport.v2.source.reader.io.exception.exception.SchemaDiscoveryException;
+import com.google.cloud.teleport.v2.spanner.migrations.schema.SourceColumnType;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import javax.sql.DataSource;
+
+/** Discover Schema of Source Tables. */
+public interface SchemaDiscovery {
```
Comment (aksharauke) [inline v2/sourcedb-to-spanner/src/main/java/com/google/cloud/teleport/v2/source/reader/io/schema/SchemaDiscovery.java:25]: Is there a use case when someone would not want to do a retry ? Can we expose only the Retriable interface and the degenerate case be no-retries?

File: v2/sourcedb-to-spanner/src/main/java/com/google/cloud/teleport/v2/source/reader/io/jdbc/dialectadapter/dialectadapter/dialectadapter/mysql/MysqlDialectAdapter.java [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+   *     reading information schema table directly, since:
+   *     <ol>
+   *       <li>With all other things equal, using a prepared statement to read information schema
+   *           would be faster than making individual calls.
+   *       <li>Various drivers have different properties that can affect the search scope of the
+   *           calls.
+   *     </ol>
+   */
+  @Override
+  public ImmutableMap<String, ImmutableMap<String, SourceColumnType>> discoverTableSchema(
```
Comment (aksharauke) [inline v2/sourcedb-to-spanner/src/main/java/com/google/cloud/teleport/v2/source/reader/io/jdbc/dialectadapter/dialectadapter/dialectadapter/mysql/MysqlDialectAdapter.java:65]: Is it in plans to also expose additional schematic information - specifically foreign key constraints?

File: v2/sourcedb-to-spanner/src/test/java/com/google/cloud/teleport/v2/source/reader/io/schema/SchemaDiscoveryImplTest.java [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+                        .withExponent(1)
+                        .withMaxRetries(testRetryCount))
+                .discoverTableSchema(mockDataSource, mockSourceSchemaReference, ImmutableList.of()))
+        .isEqualTo(ImmutableMap.of());
+    verify(mockRetriableSchemaDiscovery, times(expectedCallsCount))
+        .discoverTableSchema(any(), any(), any());
+  }
+
+  @Test
+  public void testSchemaDiscoveryImplThrowsRetiesExhausted()
```
Comment (aksharauke) [inline v2/sourcedb-to-spanner/src/test/java/com/google/cloud/teleport/v2/source/reader/io/schema/SchemaDiscoveryImplTest.java:74]: nit: Retries* in the function name

File: v2/sourcedb-to-spanner/src/main/java/com/google/cloud/teleport/v2/source/reader/io/schema/SchemaDiscovery.java (hunk 1 above)
Comment (VardhanThigle) [reply v2/sourcedb-to-spanner/src/main/java/com/google/cloud/teleport/v2/source/reader/io/schema/SchemaDiscovery.java:25]: Actually the Schema Discovery Implementation wraps the retriable interface and does the necessary retries.
The outside code will always use `SchemaDiscovery` which never throes a retriable exception.

This was done to make the retry handling logic common and avoiding the need to implement the same for say PG in future.

File: v2/sourcedb-to-spanner/src/main/java/com/google/cloud/teleport/v2/source/reader/io/jdbc/dialectadapter/dialectadapter/dialectadapter/mysql/MysqlDialectAdapter.java (hunk 2 above)
Comment (VardhanThigle) [reply v2/sourcedb-to-spanner/src/main/java/com/google/cloud/teleport/v2/source/reader/io/jdbc/dialectadapter/dialectadapter/dialectadapter/mysql/MysqlDialectAdapter.java:65]: In future yes. The dialect adapter can expose this information. Also for string splitting, we will need to provide range queries to JDBCIO (by default it always assumes an increasing order of the index), these range query prepare statements can be created by the dialect adapter.

**Review by aksharauke — APPROVED** [review]

**Pull request merged**